polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -1057,41 +1135,114 @@ mod tests {
 		let mut new_ccr = dummy_committed_candidate_receipt_v2();
 		new_ccr.descriptor.core_index = 123;
 		new_ccr.descriptor.para_id = ParaId::new(1000);
 
+		let mut cq = BTreeMap::new();
+		cq.insert(
+			CoreIndex(123),
+			vec![new_ccr.descriptor.para_id(), new_ccr.descriptor.para_id()].into(),
+		);
+		let cq = transpose_claim_queue(cq);
+
+		// No commitments
+
 		// dummy XCM messages
 		new_ccr.commitments.upward_messages.force_push(vec![0u8; 256]);
 		new_ccr.commitments.upward_messages.force_push(vec![0xff; 256]);
 
+		assert_eq!(
+			new_ccr.check_ump_signals(&cq),
+			Ok(CandidateUMPSignals { select_core: None, approved_peer: None })
+		);
+
 		// separator
 		new_ccr.commitments.upward_messages.force_push(UMP_SEPARATOR);
 
+		assert_eq!(
+			new_ccr.check_ump_signals(&cq),
+			Ok(CandidateUMPSignals { select_core: None, approved_peer: None })
+		);
+
 		// CoreIndex commitment
+		{
+			let mut new_ccr = new_ccr.clone();
+			new_ccr
+				.commitments
+				.upward_messages
+				.force_push(UMPSignal::SelectCore(CoreSelector(0), ClaimQueueOffset(1)).encode());
+
+			assert_eq!(
+				new_ccr.check_ump_signals(&cq),
+				Ok(CandidateUMPSignals {
+					select_core: Some((CoreSelector(0), ClaimQueueOffset(1))),
+					approved_peer: None
+				})
+			);
+		}
+
+		{
+			let mut new_ccr = new_ccr.clone();
+
+			// Test having only an approved peer.
+			new_ccr
+				.commitments
+				.upward_messages
+				.force_push(UMPSignal::ApprovedPeer(vec![1, 2, 3].try_into().unwrap()).encode());
+
+			assert_eq!(
+				new_ccr.check_ump_signals(&cq),
+				Ok(CandidateUMPSignals {
+					select_core: None,
+					approved_peer: Some(vec![1, 2, 3].try_into().unwrap())
+				})
+			);
+
+			// Test having an approved peer and a core selector.
+
+			new_ccr
+				.commitments
+				.upward_messages
+				.force_push(UMPSignal::SelectCore(CoreSelector(0), ClaimQueueOffset(1)).encode());
+
+			assert_eq!(
+				new_ccr.check_ump_signals(&cq),
+				Ok(CandidateUMPSignals {
+					select_core: Some((CoreSelector(0), ClaimQueueOffset(1))),
+					approved_peer: Some(vec![1, 2, 3].try_into().unwrap())
+				})
+			);
+		}
+
+		// Test having a core selector and an approved peer.
 		new_ccr
 			.commitments
 			.upward_messages
 			.force_push(UMPSignal::SelectCore(CoreSelector(0), ClaimQueueOffset(1)).encode());
+		new_ccr
+			.commitments
+			.upward_messages
+			.force_push(UMPSignal::ApprovedPeer(vec![1, 2, 3].try_into().unwrap()).encode());
 
-		let mut cq = BTreeMap::new();
-		cq.insert(
-			CoreIndex(123),
-			vec![new_ccr.descriptor.para_id(), new_ccr.descriptor.para_id()].into(),
+		assert_eq!(
+			new_ccr.check_ump_signals(&cq),
+			Ok(CandidateUMPSignals {
+				select_core: Some((CoreSelector(0), ClaimQueueOffset(1))),
+				approved_peer: Some(vec![1, 2, 3].try_into().unwrap())
+			})
 		);
-
-		assert_eq!(new_ccr.check_core_index(&transpose_claim_queue(cq)), Ok(()));
 	}
 
 	#[test]
-	fn test_invalid_ump_commitment() {
+	fn test_invalid_ump_commitments() {
```

Comment:
While also no docs. This test is well structured. I was able to get an idea of coverage quite quickly.

Reply:
yes, some tests exercise multiple cases and for these in particular I'm more of a fan of scattering comments for each case instead of having a big summary